source/common/json/json_loader.cc

```diff
@@ -2,242 +2,669 @@
+  checkType(Type::Object);
+  for (const auto& item : value_.object_value_) {
+    bool stop_iteration = !callback(item.first, *item.second);
+    if (stop_iteration) {
+      break;
     }
+  }
+}
 
-    rapidjson::SchemaDocument schema_document(document);
-    rapidjson::SchemaValidator schema_validator(schema_document);
+void Field::validateSchema(const std::string& schema) const {
+  rapidjson::Document schema_document;
+  if (schema_document.Parse<0>(schema.c_str()).HasParseError()) {
+    throw std::invalid_argument(fmt::format(
+        "Schema supplied to validateSchema is not valid JSON\n Error(offset {}) : {}\n",
+        schema_document.GetErrorOffset(), GetParseError_En(schema_document.GetParseError())));
+  }
+
+  rapidjson::SchemaDocument schema_document_for_validator(schema_document);
+  rapidjson::SchemaValidator schema_validator(schema_document_for_validator);
 
-    if (!value_.Accept(schema_validator)) {
-      // TODO(mattklein123): Improve errors by switching to SAX API.
-      rapidjson::StringBuffer schema_string_buffer;
-      rapidjson::StringBuffer document_string_buffer;
+  if (!asRapidJsonDocument().Accept(schema_validator)) {
+    rapidjson::StringBuffer schema_string_buffer;
+    rapidjson::StringBuffer document_string_buffer;
 
-      schema_validator.GetInvalidSchemaPointer().StringifyUriFragment(schema_string_buffer);
-      schema_validator.GetInvalidDocumentPointer().StringifyUriFragment(document_string_buffer);
+    schema_validator.GetInvalidSchemaPointer().StringifyUriFragment(schema_string_buffer);
+    schema_validator.GetInvalidDocumentPointer().StringifyUriFragment(document_string_buffer);
 
-      throw Exception(fmt::format(
-          "JSON object doesn't conform to schema.\n Invalid schema: {}.\n Invalid keyword: "
-          "{}.\n Invalid document key: {}",
-          schema_string_buffer.GetString(), schema_validator.GetInvalidSchemaKeyword(),
-          document_string_buffer.GetString()));
+    throw Exception(fmt::format(
+        "JSON at lines {}-{} does not conform to schema.\n Invalid schema: {}\n"
+        " Schema violation: {}\n"
+        " Offending document key: {}",
+        line_number_start_, line_number_end_, schema_string_buffer.GetString(),
+        schema_validator.GetInvalidSchemaKeyword(), document_string_buffer.GetString()));
+  }
+}
+
+bool ObjectHandler::StartObject() {
+  FieldSharedPtr object = Field::createObject();
+  object->setLineNumberStart(stream_.getLineNumber());
+
+  switch (state_) {
+  case expectValueOrStartObjectArray:
+    stack_.top()->insert(key_, object);
+    stack_.push(object);
+    state_ = expectKeyOrEndObject;
+    return true;
+  case expectArrayValueOrEndArray:
+    stack_.top()->append(object);
+    stack_.push(object);
+    state_ = expectKeyOrEndObject;
+    return true;
+  case expectRoot:
+    root_ = object;
+    stack_.push(object);
+    state_ = expectKeyOrEndObject;
+    return true;
+  default:
+    return false;
```

Comment:
per offline convo, lets do NOT_REACHED here since this should never happen and is a programming error (same elsewhere). I would add comment also on why.